### DISCUSSION TITLE
Python 3.13

### DISCUSSION THREAD

**Discussion by gusmanb** [gusmanb/logicanalyzer#153] [Announcements]
It has been detected that Python 3.13 does not work with Python.net (the library that I use to embed Python in logic analyzer).

Please, use Python 3.12 instead, once Python.net is updated to support it I will release a new version of v6.0 beta.

**Comment by opcodex64** [comment]
Hello gusmanb,

two questions:
(1) is it possible to keep python 3.12 and python3.13 in parallel
(2) if (1) is possible, where one can set the path to the python library? When LogicAnalyzer is launched, it gives the hint to edit the path to the library in the python.cfg. In the LogicAnalyzer folder there is no python.cfg. Can you give example for this or do I have here a "huge" misunderstanding?

Thank you!

Kind regards,
opcode_x64

**Reply by opcodex64** [reply]
It seems, that I could solve the issue on my system which is a MacBook Pro M3 with OSX (Sonoma 14.5), whereby I have two versions of python3 installed: 3.12 and 3.13

(1) Create a file named "python.cfg" (without quotes) in the root directory of the LogicAnalyzer Software. In other words: the python.cfg file must be located in the same directory where the LogicAnalyzer executable is located.

(2) Add the path to the library of python3 version , which is supported by Python.Net (Python 3.7 - 3.12)
However, in my case (MacOSX with brew) I have python 3.12, therefore the content of the python.cfg is the following single line: /opt/homebrew/opt/python@3.12/Frameworks/Python.framework/Versions/3.12/lib/libpython3.12.dylib

(3) Save the file ;-)

(4) Launch LogicAnalyzer Software via Terminal using: ./LogicAnalyzer -verbose
The argument  -verbose file will create (or append to) the a text file called "PythonInitLog.txt"  in the same directory. This log file will tell you (in cases) error messages, which may help you.

(Maybe these steps are also needed in your case, if you have already python3.13 installed)
Also to be mentioned, that I changed/set system globally the python3 command to dedicated version python3.12
I did this by changing symbolic link of "/usr/bin/python3" to "/opt/homebrew/bin/python3.12" via:
(1) sudo ln -s -f /usr/bin/python3 /opt/homebrew/bin/python3.12   
then executing:
(2) exec $SHELL -l

(3) To verify you can evaluate:
python3 --version

However, now I can see all decoders (sigrok) in the LogicAnalyzer Software again.

opcode_x64

**Comment by jotamudo** [comment]
For those that don't wanna mess with the python path (I didn't), you can just build the software from master using brew and update the `pythonnet` dependency on the project sln file to 3.0.5+. Works fine on my m3 air.
Don't forget to copy the decoders/ folder to the built directory as it's the one it searches when booting

**Comment by mz1874** [comment]
3.12 still not works

2025/3/1 8:15 -> Initializing python system...
2025/3/1 8:15 -> Reading python path from config file...
2025/3/1 8:15 -> Stablished path: C:\Users\23391\AppData\Local\Programs\Python\Python312\python312.dll
2025/3/1 8:15 -> Initializing decoders...
2025/3/1 8:15 -> Error initializing python engine: The type initializer for 'Delegates' threw an exception. -    at Python.Runtime.Runtime.Delegates.get_Py_IsInitialized()
   at Python.Runtime.Runtime.Py_IsInitialized()
   at Python.Runtime.Runtime.<>c.<Initialize>b__32_0()
   at Python.Runtime.Runtime.TryUsingDll[T](Func`1 op)
   at Python.Runtime.Runtime.Initialize(Boolean initSigs)
   at Python.Runtime.PythonEngine.Initialize(IEnumerable`1 args, Boolean setSysArgv, Boolean initSigs)
   at Python.Runtime.PythonEngine.Initialize(Boolean setSysArgv, Boolean initSigs)
   at Python.Runtime.PythonEngine.Initialize()
   at SigrokDecoderBridge.SigrokPythonEngine.EnsureInitialized() in F:\LogicAnalyzerRepo\Software\LogicAnalyzer\LogicAnalyzer\SigrokDecoderBridge\SigrokPythonEngine.cs:line 99

**Comment by physiker66** [comment]
I found 3 reasons for this error (all of them involving me being stupid... but maybe it helps someone else):
a) not setting the full path to pythonXXX.dll in python.cfg, instead entering the directory
b) 32bit python vs 64bit application
c) using python 3.12 (3.11 works)

**Comment by arekm** [comment]
"Python.NET is currently compatible and tested with Python releases 3.7 - 3.13." where Python.NET is >= [v3.0.5](https://github.com/pythonnet/pythonnet/releases/tag/v3.0.5)

**Reply by gusmanb** [reply]
This is old, python 3.13 is already supported by the analyzer, I will close the discussion to avoid any confusion.

**Comment by amadeok** [comment]
I think the issue i had is that the python310.dll had a p lowercase, while the program was looking for Python310.dll, p uppercase
